### ISSUE TITLE
Lints not showing in pubspec files.

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#35168]
Possibly mis-filed on the IDEA issue tracker: https://youtrack.jetbrains.com/issue/WEB-35763, DAS appears to, in at least some cases, not be reporting analysis errors on pubspecs.

Here's a repro.

1. create a Dart project
2. create an `analysis_options.yaml` file w/ the `package_names` lint enabled:
```
linter:
  rules:
    - package_names
```

3. update the package name in the pubspec to something camelcase (e.g., FooBar) and save

**What is the expected result?**

Should see an error reported in the pubspec.

**What happens instead?**

No reported error.

/cc @bwilkerson @stereotype441 

See also: #31890.

**Comment by srawlins** [comment]
This is still a bug for me; lint does not show up in Problems panel of VS Code until I save the pubspec.yaml change.

**Comment by bwilkerson** [comment]
Is server sent an overlay notification for `pubspec.yaml` files? If so, do the updated diagnostics get sent to the client? (Just trying to figure out where the bug is.)

**Comment by pq** [comment]
> Is server sent an overlay notification for pubspec.yaml files? 

Maybe @jwren knows?

**Comment by alexander-doroshko** [comment]
AFAIR, reporting diagnostics only for the *saved* state of the file `pubspec.yaml` or `analysis_options.yaml` file was a [known limitation](https://github.com/dart-lang/sdk/issues/25114#issuecomment-697397940). Is Analysis Server now expected to handle unsaved files as well?

**Comment by bwilkerson** [comment]
I don't think so, I was just questioning my assumptions. The reason we decided against that is because changes to either `pubspec.yaml` or `analysis_options.yaml` would necessitate re-analyzing the whole package, and that would be too expensive.

If we want to change that decision then we'll need to figure out the semantic of what it means for the content to be overlayed. Maybe it would be as simple as "re-analyze the YAML file, but don't reanalyze anything else". That could still result in some surprising behavior from a user's perspective, though so I'm not convinced it would be a good change. (For example, adding a lint rule wouldn't run the lint until the file was saved, even though it looks to the user like it's been enabled.)

In any case, it looks like it's currently working as intended.

**Comment by DanTup** [comment]
> Maybe it would be as simple as "re-analyze the YAML file, but don't reanalyze anything else".

I think that behaviour would make sense. I generally expect the IDE to give me up-to-date feedback on things that are valid/invalid. I don't think that necessarily means that unsaved files should be immediately "applied" where they would trigger actions (we already have this distinction - for example I can fix Dart analysis errors and the squiggles disappear, but I may still need to save for a hot reload or to run the code).


> For example, adding a lint rule wouldn't run the lint until the file was saved, even though it looks to the user like it's been enabled

I'm not sure what "looks like it's been enabled" means here though - doesn't the file look the same regardless of whether the lint is applied/saved or not?

**Comment by bwilkerson** [comment]
> I'm not sure what "looks like it's been enabled" means here though - doesn't the file look the same regardless of whether the lint is applied/saved or not?

Yes, which is what I was trying to say. The indication that the file hasn't been saved is subtle enough that I suspect some users won't understand why the lint isn't being applied despite the name clearly being in the list. But you might be right in thinking that the distinction between written and saved is common enough that it would rarely be a problem.

I'm going to decrease the priority of this issue because of the lack of feedback that this is a significant problem for users, but leave it open because it might be a nice improvement in some cases.

**Comment by srawlins** [comment]
See, I believe, the sibling issue for analysis options here: https://github.com/dart-lang/sdk/issues/52424
